drivers/nvdimm/pmem.c

```diff
@@ -45,11 +65,352 @@ static struct nd_region *to_region(struct pmem_device *pmem)
+	struct device *dev = to_dev(tr->pmem);
+
+	if (result->result != DMA_TRANS_NOERROR) {
+		dev_err(dev, "DMA error %x\n", result->result);
+		if (result->result == DMA_TRANS_ABORTED)
+			bio->bi_status = BLK_STS_TRANSPORT;
+		else
+			bio->bi_status = BLK_STS_IOERR;
+	}
+	if (refcount_dec_and_test(&tr->inprog)) {
+		if (tr->do_acct)
+			bio_end_io_acct(bio, tr->start);
+		pmem_dma_unmap(tr);
+		bio_endio(bio);
+	}
+}
+
+static bool pmem_dma_submit_bio(struct pmem_device *pmem, struct bio *bio,
+    bool do_acct, unsigned long start)
+{
+	struct device *dev = to_dev(pmem);
+	struct dma_chan	*dma_chan1, *dma_chan2;
+	struct dma_device *dma_dev1, *dma_dev2;
+	struct dma_async_tx_descriptor *tx;
+	dma_async_tx_callback_result cb;
+	enum dma_data_direction dir;
+	dma_cookie_t cookie, last_cookie1 = 0, last_cookie2 = 0;
+	struct pmem_dma_tr *tr;
+	phys_addr_t laddr, raddr;
+	struct bio_vec bvec;
+	struct bvec_iter iter;
+	unsigned long flags;
+	unsigned int dmas, i, vecs = 0;
+	sector_t sector;
+	static struct dmaengine_result dummy_result = {
+		.result = DMA_TRANS_ABORTED,
+		.residue = 0
+	};
+
+	/* For small I/Os softwate copy is faster. */
+	if (bio->bi_iter.bi_size < min_dma_size)
+		return false;
+
+	/* Choose DMA channels for local and remote PMEM accesses. */
+	laddr = pmem->phys_addr + pmem->data_offset;
+	raddr = op_is_write(bio_op(bio)) ? pmem->rphys_addr : 0;
+	if (raddr)
+		raddr += pmem->data_offset;
+	if (!raddr && pmem->rdma_for_single) {
+		dma_chan1 = pmem->rdma_chan;
+		dma_chan2 = pmem->dma_chan;
+	} else {
+		dma_chan1 = pmem->dma_chan;
+		dma_chan2 = pmem->rdma_chan;
+	}
+	if (dma_chan1 == NULL || dma_chan2 == NULL)
+		return false;
+
+	/* Check the addresses alignment fit DMA device(s) requirements. */
+	dma_dev1 = dma_chan1->device;
+	dma_dev2 = dma_chan2->device;
+	bio_for_each_bvec(bvec, bio, iter) {
+		if (!is_dma_copy_aligned(dma_dev1, bvec.bv_offset,
+		    laddr + iter.bi_sector * 512, bvec.bv_len))
+			return false;
+		if (raddr &&
+		    !is_dma_copy_aligned(dma_dev2, bvec.bv_offset,
+		    raddr + iter.bi_sector * 512, bvec.bv_len))
+			return false;
+		vecs++;
+	}
+
+	/* Collect information needed to complete BIO on DMA completion. */
+	dmas = raddr ? 2 : 1;
+	tr = kmalloc(offsetof(struct pmem_dma_tr, addr[vecs * dmas]),
+	    GFP_NOWAIT | __GFP_ZERO);
+	if (!tr) {
+		dev_warn(dev, "kmalloc() failed\n");
+		return false;
+	}
+	tr->pmem = pmem;
+	tr->bio = bio;
+	tr->single = !raddr;
+	tr->dma_dev1 = dma_dev1->dev;
+	tr->dma_dev2 = dma_dev2->dev;
+	refcount_set(&tr->inprog, dmas);
+	tr->do_acct = do_acct;
+	tr->start = start;
+
+	/* Map local PMEM for the first DMA device. */
+	sector = bio->bi_iter.bi_sector;
+	dir = op_is_write(bio_op(bio)) ? DMA_FROM_DEVICE : DMA_TO_DEVICE;
+	tr->laddr = dma_map_resource(dma_dev1->dev,
+	    laddr + sector * 512, bio->bi_iter.bi_size, dir, 0);
+	if (dma_mapping_error(dma_dev1->dev, tr->laddr)) {
+		dev_warn(dev, "dma_map_page() 1 failed\n");
+		pmem_dma_unmap(tr);
```

Comment:
We can free ```tr``` also in the below conditions?

Reply:
I am not getting this question.  pmem_dma_unmap(tr) also frees the tr.  And we need to do it in every case where we return false.  When we return true it will be called by the DMA callback.  If you mean that I could just call kfree(tr) in this particular case -- yes I could, but preferred uniformity.

Reply:
Ah, I see. I was just thinking about where memory is freed, Thank you for the explanation.